test/server/active_internal_listener_test.cc

```diff
@@ -0,0 +1,230 @@
+      new Network::Address::Ipv4Instance("127.0.0.2", 8080));
+
+  Network::MockConnectionSocket* accepted_socket = new NiceMock<Network::MockConnectionSocket>();
+
+  EXPECT_CALL(filter_chain_factory_, createListenerFilterChain(_))
+      .WillRepeatedly(Invoke([&](Network::ListenerFilterManager& manager) -> bool {
+        // Insert the Mock filter.
+        manager.addAcceptFilter(listener_filter_matcher_,
+                                Network::ListenerFilterPtr{test_listener_filter});
+        return true;
+      }));
+  EXPECT_CALL(*test_listener_filter, onAccept(_))
+      .WillOnce(Invoke([&](Network::ListenerFilterCallbacks& cb) -> Network::FilterStatus {
+        cb.socket().connectionInfoProvider().restoreLocalAddress(original_dst_address);
+        return Network::FilterStatus::Continue;
+      }));
+  EXPECT_CALL(*test_listener_filter, destroy_());
+  auto filter_factory_callback = std::make_shared<std::vector<Network::FilterFactoryCb>>();
+  filter_chain_ = std::make_shared<NiceMock<Network::MockFilterChain>>();
+  auto transport_socket_factory = Network::Test::createRawBufferSocketFactory();
+
+  EXPECT_CALL(manager_, findFilterChain(_)).WillOnce(Return(filter_chain_.get()));
+  EXPECT_CALL(*filter_chain_, transportSocketFactory)
+      .WillOnce(testing::ReturnRef(*transport_socket_factory));
+  EXPECT_CALL(*filter_chain_, networkFilterFactories).WillOnce(ReturnRef(*filter_factory_callback));
+  auto* connection = new NiceMock<Network::MockServerConnection>();
+  EXPECT_CALL(dispatcher_, createServerConnection_()).WillOnce(Return(connection));
+  EXPECT_CALL(conn_handler_, incNumConnections());
+  EXPECT_CALL(filter_chain_factory_, createNetworkFilterChain(_, _)).WillOnce(Return(true));
+  EXPECT_CALL(listener_config_, perConnectionBufferLimitBytes());
+  internal_listener_->onAccept(Network::ConnectionSocketPtr{accepted_socket});
+  EXPECT_CALL(conn_handler_, decNumConnections());
+  connection->close(Network::ConnectionCloseType::NoFlush);
+  dispatcher_.clearDeferredDeleteList();
+  EXPECT_CALL(*generic_listener_, onDestroy());
+}
+
+TEST_F(ActiveInternalListenerTest, StopListener) {
+  addListener();
+  EXPECT_CALL(*generic_listener_, onDestroy());
+  internal_listener_->shutdownListener();
+}
+
+TEST_F(ActiveInternalListenerTest, PausedListenerAcceptNewSocket) {
+  addListenerWithRealNetworkListener();
+  internal_listener_->pauseListening();
+
+  expectFilterChainFactory();
+  Network::MockConnectionSocket* accepted_socket = new NiceMock<Network::MockConnectionSocket>();
+
+  EXPECT_CALL(filter_chain_factory_, createListenerFilterChain(_))
+      .WillRepeatedly(Invoke([&](Network::ListenerFilterManager&) -> bool { return true; }));
+  EXPECT_CALL(manager_, findFilterChain(_)).WillOnce(Return(nullptr));
+  internal_listener_->onAccept(Network::ConnectionSocketPtr{accepted_socket});
+}
+
+TEST_F(ActiveInternalListenerTest, DestroyListenerCloseAllConnections) {
```

Comment:
Is there a notion of draining an internal listener?

Reply:
Sorry, I don't fully get your question.
Our usage of the internal listener includes the cases each internal listener(or filter chain within) represents a L7 service. Since the l7 service could be http, it's nice to close the connection to internal listener only after a certain timeout.

Reply:
Conceptually drain is only applicable to downstream listeners that accept connections from external clients. Internal listener accepts connections from upstream connections only. As a result I do not think drain would be applicable.

Reply:
Chat with @yanavlasov  offline. Drain is derived from internal listener update and it's good to gracefully close the connections as in this test case.